components/speedreader/speedreader_component.h

```diff
@@ -0,0 +1,68 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_SPEEDREADER_SPEEDREADER_COMPONENT_H_
+#define BRAVE_COMPONENTS_SPEEDREADER_SPEEDREADER_COMPONENT_H_
+
+#include <memory>
+#include <string>
+
+#include "base/files/file_path_watcher.h"
+#include "base/memory/weak_ptr.h"
+#include "base/observer_list.h"
+#include "brave/components/brave_component_updater/browser/brave_component.h"
+
+namespace base {
+class FilePath;
+}
+
+class GURL;
+
+namespace speedreader {
+
+class SpeedreaderComponent
+    : public brave_component_updater::BraveComponent,
+      public base::SupportsWeakPtr<SpeedreaderComponent> {
+ public:
+  class Observer : public base::CheckedObserver {
+   public:
+    virtual void OnWhitelistReady(const base::FilePath& path) = 0;
+    virtual void OnStylesheetReady(const base::FilePath& path) = 0;
+
+   protected:
+    ~Observer() override = default;
+  };
+
+  explicit SpeedreaderComponent(Delegate* delegate);
+  ~SpeedreaderComponent() override;
+
+  SpeedreaderComponent(const SpeedreaderComponent&) = delete;
+  SpeedreaderComponent& operator=(const SpeedreaderComponent&) = delete;
+
+  void AddObserver(Observer* observer);
+  void RemoveObserver(Observer* observer);
+
+  const base::FilePath& GetWhitelistPath() { return whitelist_path_; }
+  const base::FilePath& GetStylesheetPath() { return stylesheet_path_; }
+
+ private:
+  // brave_component_updater::BraveComponent:
+  void OnComponentReady(const std::string& component_id,
+                        const base::FilePath& install_dir,
+                        const std::string& manifest) override;
+
+  // Used in testing/development with custom rule set for auto-reloading
+  void OnWhitelistFileReady(const base::FilePath& path, bool error);
+
+  base::ObserverList<Observer> observers_;
```

Comment:
I think we can simplify by just holding a raw pointer, given that the only observer by design is an actual owner

Reply:
You actually can pass this pointer to Component constructor

Reply:
thought of that, but went for a solution that better abstracts away the specifics of speedreader and attempts to follow a more decoupled pattern of using component updater.

Happy to couple them up more if you prefer?

Perhaps might be even better to initialise `SpeedreaderComponent` outside of the `SpeedreaderRewriterService` altogether and just subscribe the service? Didn't want to add too much to the browser process either